### PR TITLE
[AutoDiff] NFC: reorganize differentiation SILOptimizer files.

### PR DESCRIPTION
**Pull request by dan-zheng**
Move differentiation-related SILOptimizer files to `{include/swift,lib}/SILOptimizer/Differentiation/`.

This reduces directory nesting within `lib/SILOptimizer/` and gathers files together.

Requested by @gottesmm.

---

Further cleanup:
-  TF-1218: Use the pointer-to-impl pattern for `[JVP,VJP,Pullback]Emitter`. This should minimize their header files to just `run()` entry points, eliminating duplication with `.cpp` files.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@gottesmm: could you please review, as a SILOptimizer code owner?
You asked for reorganization: https://github.com/apple/swift/pull/30818#pullrequestreview-391698895

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@gottesmm has more feedback on organization. We're discussing at swift-devel.slack.com on the `#autodiff` channel now.

File: include/swift/SILOptimizer/Differentiation/PullbackEmitter.h [hunk 1]
```diff
@@ -18,11 +18,12 @@
 #ifndef SWIFT_SILOPTIMIZER_UTILS_DIFFERENTIATION_PULLBACKEMITTER_H
 #define SWIFT_SILOPTIMIZER_UTILS_DIFFERENTIATION_PULLBACKEMITTER_H
 
-#include "swift/SIL/TypeSubstCloner.h"
+#include "AdjointValue.h"
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Differentiation/PullbackEmitter.h:21]: You should include these with full paths. We generally do that in headers.

Comment (dan-zheng) [reply include/swift/SILOptimizer/Differentiation/PullbackEmitter.h]: Done. I'll keep that in mind.

**Review by gottesmm — APPROVED** [review]
LGTM

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Pull request merged**